### PR TITLE
Better handling of Wasm Names ##bin

### PR DESCRIPTION
**Pull request by swoops**
- [ ] Mark this if you consider it ready to merge
- [ ] I've added tests (optional)
- [ ] I wrote some lines in the [book](https://github.com/radareorg/radare2book) (optional)

**Description**

A Wasm "Name" is valid if it's utf-8, which can include NULL bytes and all kind of other crazy things. So to prevent r2 from choking on this later in tcc or something, I encode the names on entry. This means we don't have to keep the string length and we can decode them later on output if we want.

Encoder might be a bit overkill atm, I pretty much encode every WASM name into something that can be treated as a C var. I probably don't need to encode section names so strongly, for example I could let a section name start with a number or `.`. So maybe this is just a stop gap.

Feel free to ask for better encoding or accept for now.

### CODE REVIEW

File: libr/bin/format/wasm/wasm.c [hunk 1]
```diff
@@ -105,6 +105,60 @@ static bool consume_str_new(RBuffer *b, ut64 bound, ut32 *len_out, char **str_ou
...
+		free (orig);
+		return false;
+	}
+
+	size_t i, oi = 0;
+	for (i = 0; i < len; i++) {
+		if (WASM_IS_OK (orig, i, len)) {
+			sout[oi++] = orig[i];
+		} else {
+			oi += sprintf (sout + oi, "_%02x_", orig[i]);
```
Comment (trufae) [inline libr/bin/format/wasm/wasm.c:143]: use snprintf passingn len - oi as size, maybe check for overflow may be good in the other case

Comment (swoops) [reply libr/bin/format/wasm/wasm.c]: Should be fixed, along with a pointer mistake

Comment (swoops) [reply libr/bin/format/wasm/wasm.c]: Another fix... I should use length of sout rather then length of orig.

File: libr/bin/format/wasm/wasm.c [hunk 2]
```diff
@@ -105,6 +105,61 @@ static bool consume_str_new(RBuffer *b, ut64 bound, ut32 *len_out, char **str_ou
...
+#define WASM_IS_OK(str, i, max) (WASM_IS_OK_EXCEPT__ (i, str[i]) || WASM_UNDERSCORE_OK (str, i, max))
+static bool consume_encoded_name_new(RBuffer *b, ut64 bound, ut32 *len_out, char **str_out) {
+	ut32 len;
+	char *orig = NULL;
+	if (!consume_str_new (b, bound, &len, &orig)) {
+		return false;
+	}
+
+	// room for even every character getting encoded
+	ut32 maxsize = len * 4 + 2;
```
Comment (trufae) [inline libr/bin/format/wasm/wasm.c:132]: size_t
```suggestion
	size_t maxsize = len * 4 + 2;
```

File: libr/bin/format/wasm/wasm.c [hunk 3]
```diff
@@ -105,6 +105,61 @@ static bool consume_str_new(RBuffer *b, ut64 bound, ut32 *len_out, char **str_ou
...
+	char *sout = malloc (maxsize);
+	if (!sout) {
+		free (orig);
+		return false;
+	}
+
+	size_t i, oi = 0;
+	for (i = 0; i < len; i++) {
+		if (WASM_IS_OK (orig, i, len)) {
```
Comment (trufae) [inline libr/bin/format/wasm/wasm.c:141]: Add check for oi +4 < maxsize break

**Pull request merged**